### ISSUE TITLE
Why are Telemetry/snapper/Spesh/MoarVM::Profiler core modules?

### ISSUE THREAD

**Issue by ugexe** [rakudo/rakudo#3161]
## The Problem

Core modules should be reserved for specific circumstances. In the Perl5 world it is generally reserved for modules required to build and install other modules. For Perl6 Test is used for testing, and NativeCall is used for checking if native libraries are installed. Telemetry/snapper/Spesh/MoarVM::Profiler are not used for any build/install related purposes, not declared as a dependency of anything, yet still impose versioning issues unique to core modules.

## Expected Behavior

Telemetry/snapper/Spesh/MoarVM::Profiler are moved into ecosystem modules. (Pod::To::Text I'm not sure, but it feels like something a build system might use for installing documentation?)

(side note: CompUnit::Repository::Staging should be moved from `lib/` to `core/` -- that might still not be possible and is the reason it is in `lib/` instead of `core/` to begin with)

## Actual Behavior

Some modules are added to core seemingly based on their coupling to some non-rakudo component or unofficial API, despite it being inevitable (and current reality) that ecosystems will contain modules that only work on specific backends anyway. As for adding them for usefulness -- I'm sure most people would rather have a core http client (for better or worse) out-of-the-box than the ability to profile out-of-the-box, yet we do not have a core http client.

**Comment by jnthn** [comment]
I've not gone though all the modules mentioned, and this may only apply to some of them, but...

> Some modules are added to core seemingly based on their coupling to some non-rakudo component or unofficial API

Depending on the nature of the coupling (especially if it's use of nqp::ops), this feels like a sensible reason to me. If we want to change, for example, the internals of how NativeCall things work, it's vastly easier to do that if it's installed along with the compiler, which in turn depends on a minimum MoarVM version. For things which expose a similar runtime binding, it probably makes sense to keep them alongside the compiler.

Glancing at some of those in question, `Spesh` feels like something that should probably live on the ecosystem, but `Telemetry` at least starts with `use nqp` and `MoarVM::Profiler` couples to the VM-provided profiling ops.

> despite it being inevitable (and current reality) that ecosystems will contain modules that only work on specific backends anyway

Perhaps so, but even in cases where the change we're making is indisputably a bug fix, there's still pressure to not break the ecosystem. Now imagine I have to argue something like, "I've broken this module because I wanted to change the API of the internal thing to something more flexible/performant, and can't afford the time to worry about back-compat". I'd rather not need to have that discussion at all.

In short, there's no incentive for those of us doing compiler/runtime dev to have things coupling tightly to the runtime being more costly to change. For cases where that argument doesn't apply, sure, let's move them.

**Comment by lizmat** [comment]
- Telemetry / snapper: use a lot of `nqp` ops, mostly for speed as to interfere the least with what they are trying to measure, but also for specific features (e.g. `nqp::getrusage`).  And the heap snapshot capability is definitely tightly bound to Rakudo / MoarVM.

- Spesh: spesh is a feature of the Rakudo / MoarVM implementation, therefore tightly bound I would say.  One could argue this should be renamed to `MoarVM::Spesh`

- MoarVM::Profiler: definitely tightly bound to MoarVM.

Spesh / MoarVM::Profiler are works in progress.  I hope to find the energy to work on these again after the whole renaming event has passed.
